x/btcstaking/keeper/power_dist_change.go

```diff
@@ -170,19 +170,28 @@ func (k Keeper) ProcessAllPowerDistUpdateEvents(
 		switch typedEvent := event.Ev.(type) {
 		case *types.EventPowerDistUpdate_BtcDelStateUpdate:
 			delEvent := typedEvent.BtcDelStateUpdate
+			btcDel, err := k.GetBTCDelegation(ctx, delEvent.StakingTxHash)
+			if err != nil {
+				panic(err) // only programming error
+			}
 			if delEvent.NewState == types.BTCDelegationStatus_ACTIVE {
 				// newly active BTC delegation
-				btcDel, err := k.GetBTCDelegation(ctx, delEvent.StakingTxHash)
-				if err != nil {
-					panic(err) // only programming error
-				}
+
 				// add the BTC delegation to each restaked finality provider
 				for _, fpBTCPK := range btcDel.FpBtcPkList {
 					fpBTCPKHex := fpBTCPK.MarshalHex()
 					activeBTCDels[fpBTCPKHex] = append(activeBTCDels[fpBTCPKHex], btcDel)
 				}
 			} else if delEvent.NewState == types.BTCDelegationStatus_UNBONDED {
-				// add the expired BTC delegation to the map
+				sdkCtx := sdk.UnwrapSDKContext(ctx)
+				// delegation expired and become unbonded emit block event about this
+				// information
+				if !btcDel.IsUnbondedEarly() {
+					if err := sdkCtx.EventManager().EmitTypedEvent(delEvent); err != nil {
+						panic(fmt.Errorf("failed to emit EventBTCDelegationStateUpdate for the new unbonded BTC delegation: %w", err))
+					}
+				}
```

Comment:
Can we emit this inside `BTCUndelegate`? A slight concern is that this couples emitting events with the main responsibility of this function, which is computing voting power table.

Later on if voting power table is moved to `x/finality` then we could not emit staking events here

Reply:
Hmm I just realised that this is the event of timelock unbonding but not early unbonding :sweat_smile: sorry for the confusion. Then yeah this is the best place to emit this event imo